rclpy/src/rclpy/action_client.cpp

```diff
@@ -0,0 +1,335 @@
+  const rmw_qos_profile_t & cancel_service_qos,
+  const rmw_qos_profile_t & feedback_topic_qos,
+  const rmw_qos_profile_t & status_topic_qos)
+: node_handle_(std::make_shared<Handle>(pynode))
+{
+  auto node = node_handle_->cast<rcl_node_t *>("rcl_node_t");
+
+  rosidl_action_type_support_t * ts =
+    static_cast<rosidl_action_type_support_t *>(rclpy_common_get_type_support(
+      pyaction_type.ptr()));
+  if (!ts) {
+    throw py::error_already_set();
+  }
+
+  rcl_action_client_options_t action_client_ops = rcl_action_client_get_default_options();
+
+  action_client_ops.goal_service_qos = goal_service_qos;
+  action_client_ops.result_service_qos = result_service_qos;
+  action_client_ops.cancel_service_qos = cancel_service_qos;
+  action_client_ops.feedback_topic_qos = feedback_topic_qos;
+  action_client_ops.status_topic_qos = status_topic_qos;
+
+  rcl_action_client_ = std::shared_ptr<rcl_action_client_t>(
+    new rcl_action_client_t,
+    [this](rcl_action_client_t * action_client)
+    {
+      auto node = node_handle_->cast_or_warn<rcl_node_t *>("rcl_node_t");
+
+      rcl_ret_t ret = rcl_action_client_fini(action_client, node);
+      if (RCL_RET_OK != ret) {
+        // Warning should use line number of the current stack frame
+        int stack_level = 1;
+        PyErr_WarnFormat(
+          PyExc_RuntimeWarning, stack_level, "Failed to fini publisher: %s",
+          rcl_get_error_string().str);
+        rcl_reset_error();
+      }
+      delete action_client;
+    });
+
+  *rcl_action_client_ = rcl_action_get_zero_initialized_client();
+
+  rcl_ret_t ret = rcl_action_client_init(
+    rcl_action_client_.get(),
+    node,
+    ts,
+    action_name,
+    &action_client_ops);
+  if (RCL_RET_ACTION_NAME_INVALID == ret) {
+    std::string error_text{"Failed to create action client due to invalid topic name '"};
+    error_text += action_name;
+    error_text += "' : ";
+    error_text += rcl_get_error_string().str;
+    rcl_reset_error();
+    throw py::value_error(error_text);
+  }
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to create action client: "};
+    error_text += rcl_get_error_string().str;
+    rcl_reset_error();
+    throw py::value_error(error_text);
+  }
+}
+
+#define TAKE_SERVICE_RESPONSE(Type) \
+  /* taken_msg is always destroyed in this function */ \
+  auto taken_msg = create_from_py(pymsg_type); \
+  rmw_request_id_t header; \
+  rcl_ret_t ret = rcl_action_take_ ## Type ## _response( \
+    rcl_action_client_.get(), &header, taken_msg.get()); \
+  int64_t sequence = header.sequence_number; \
+  /* Create the tuple to return */ \
+  if (RCL_RET_ACTION_CLIENT_TAKE_FAILED == ret || RCL_RET_ACTION_SERVER_TAKE_FAILED == ret) { \
+    return py::make_tuple(py::none(), py::none()); \
+  } else if (RCL_RET_OK != ret) { \
+    throw rclpy::RCLError("Failed to take " #Type); \
+  } \
+  return py::make_tuple(py::int_(sequence), convert_to_py(taken_msg.get(), pymsg_type)); \
```

Comment:
@ahcorde nit^N: I think you can just pass `sequence` and `pybind11` will do the magic:
```suggestion
  return py::make_tuple(sequence, convert_to_py(taken_msg.get(), pymsg_type)); \
```